### PR TITLE
add 4 multiply instructions and tests

### LINKED ISSUE #300: Multiply instructions MGRK, MSGRKC, MSRKC not doing anything
in pz390 supporting code for MGRK (x'B9EC'), MSGRKC (X'B9ED'), MSRKC (X'B9FD') are implemented as a no-op.
Please add code to execute the multiplication as described in the PoP.

### PR DESCRIPTION
**Pull request by dsh33782**
Add support for MGRK, MSGRKC, MSRKC, MSC and add regression tests.

### CODE REVIEW

**Comment by adelosa** [comment]
Hi @dsh33782 - I have created a new branch on the main repo that resolves the conflicts.. See https://github.com/z390development/z390/tree/%23300MPY. But I think you should still try the merge yourself as per the instructions on google groups. If you get stuck, let me know.

**Comment by jyganci** [comment]
Anthony,

How did you do this? Should we all know how to do this?

I agree that we should all know how to tackle merge conflicts for our pull
requests.

**Comment by dsh33782** [comment]
Anthony

 

I tried following the steps you outlined and have ended up losing the changes I made locally.  I will restore from backup, but please go ahead and try to merge the changes and complete pull for #300MPY.  I did test all the changes locally.

 

I need more hands on training with git.

 

Don Higgins

 ***@***.***> ***@***.***

www.don-higgins.net

**Comment by jyganci** [comment]
Don, Anthony,

I started looking over the pull request yesterday, even though it has merge
conflicts. During my preliminary review, some questions arose. I decided to
back off doing a review until the merge conflicts are resolved.

If Anthony fixes the merge conflicts, who will be responsible for acting on
any possible issues arising from a review of the pull request? Don's fork
and clone will be out of sync, correct?

Regards,

John Ganci

**Review by abekornelis — DISMISSED** [review]

**Comment by abekornelis** [comment]
I glossed over the changes - I am not nearly as meticulous as John ;-)
Looks okay to me - I'm not aware of any overt conflict with other pending changes.

**Comment by jyganci** [comment]
I've found typos and errors while scanning through the pull request. When
the
pull request does not have any merge conflicts I will add my comments.

**Comment by adelosa** [comment]
I used the GitHub online interface to resolve the conflict. Please review again as the merge cancels the approval.

File: src/pz390.java [hunk 1]
```diff
@@ -390,7 +390,8 @@
      * 2021-10-29 Issue 305. Fix XHEXI logic so that it matches ASSIST documentation.
      *                       NB: RPI 878 added ASCII support for XDECI but not XHEXI.
      *                           Not added here either.
-     * 2022-02-12 Issue 195. For all privileged instructions issue the proper exception according PoP
+	 * 2022-02-24 DSH ISSUE 330 add support for MGRK, MSGRKC, MSRKC, MSC
+   * 2022-02-12 Issue 195. For all privileged instructions issue the proper exception according PoP
 	 *                       Being S0C2 for privileged-operation exception when executed in problem state
 	 *                       Other exceptions as defined in PoP
 	 *                       And implement as a no-op if all tests are passed
```
Comment (jyganci) [inline src/pz390.java]: "ISSUE 330 add ..." should be "Issue 300 add ...". Correct?

A nit, but the 2022-02-24 comment should be below the 20220212 comment.

File: src/pz390.java [hunk 2]
```diff
@@ -8233,7 +8256,20 @@ private void ins_B9XX(){
...
+				  if (rlv1 != (long)rv1){
+					  psw_cc = psw_cc3;
+				  } else if (rv1 > 0){
+					  psw_cc = psw_cc2;
+				  } else if (rv1 == 0) {
+					  psw_cc = psw_cc0;
+				  } else {
+                      psw_cc = psw_cc1;
+				  }
 	    	     break;
```
Comment (jyganci) [inline src/pz390.java:8281]: See MSGRKC comment about setting the condition code and possible fixed-point overflow.

File: src/pz390.java [hunk 3]
```diff
@@ -9233,6 +9269,22 @@ private void ins_E3XX(){   // RPI 1207
...
+			if (rlv1 != (long)rv1){
+			  psw_cc = psw_cc3;
+			} else if (rv1 > 0){
+			  psw_cc = psw_cc2;
+			} else if (rv1 == 0) {
+			  psw_cc = psw_cc0;
+			} else {
+              psw_cc = psw_cc1;
+			}
+		    break;
```
Comment (jyganci) [inline src/pz390.java]: See MSGRKC comment about setting the condition code and possible fixed-point overflow.

File: src/pz390.java [hunk 4]
```diff
@@ -9436,23 +9488,32 @@ private void ins_E3XX(){   // RPI 1207
...
+			         .multiply(big_int2);
+			rlv1 = big_int1.longValue();
+			fp_bi_to_wreg(reg_byte,rf1,big_int1, 8);
+			if (big_int1.compareTo(bi_max_pos_long) == 1
+			|| big_int1.compareTo(bi_min_neg_long) == -1) {                  
+                psw_cc = psw_cc3; 
+			} else if (rlv1 > 0){
+				psw_cc = psw_cc2;
+			} else if (rlv1 == 0) {
+				psw_cc = psw_cc0;
```
Comment (jyganci) [inline src/pz390.java:9504]: See MSGRKC comment about setting the condition code and possible fixed-point overflow.

File: src/pz390.java [hunk 5]
```diff
@@ -9436,23 +9488,32 @@ private void ins_E3XX(){   // RPI 1207
...
 			} else {
-				psw_cc = get_long_comp_cc(rlv1, 0);
-			}
-			break;
+				psw_cc = psw_cc1;
+			}               
+ 			break;
```
Comment (jyganci) [inline src/pz390.java:9516]: See MSGRKC comment about setting the condition code and possible fixed-point overflow.

File: tests/TESTINS5.MLC [hunk 6]
```diff
@@ -341,6 +373,146 @@ BIC_OK2  EQU    *
...
+         CL     R1,=F'-6'
+         RT1  CC0
+         L       R1,=F'-2'
+         MSY    R1,=F'-3'
+          CL     R1,=F'6'
+          RT1  CC0
+* MSGRKC
+         LMG   R5,R6,=FL8'2,3'
+         MSGRKC  R2,R5,R6
+         CLG    R2,=FL8'6'
```
Comment (jyganci) [inline tests/TESTINS5.MLC]: Since MSGRKC sets the condition code, which would be 2 here, I think, you should first set the condition code to something other than the expected condition code then test it after the instruction executes. Something like
```
         LMG   R5,R6,=FL8'2,3'
         SR    R0,R0      Set condition code to 0
         MSGRKC  R2,R5,R6
         RT1   CC2        Condition code must be 2
         CLG   R2,=FL8'6'
```
There should be at least one test for each condition code (0,1,2,3). 

Somewhere there needs to be tests for fixed-point overflow. The technique used in TESTINS2.MLC uses ESPIE (and ESTAE) to trap program interrupt code 8, then test an instruction that results in a fixed-point overflow. For example, see TESTINS2.MLC lines 637-632 where the AR instruction is tested.

Similar code for all the MSC, MSCG, MSGRKC, MSRKC tests that follow.

File: src/pz390.java [hunk 7]
```diff
@@ -13875,14 +13936,28 @@ private int get_long_log_sub_cc() {
...
 		}
 	}
-
-	private byte[] get_log_bytes(byte[] data_byte, int data_offset, int data_len) {
+	private byte[] get_unsigned_bytes(byte[] data_byte, int data_offset, int data_len) {  // #300
+		/*
+		 * return byte array with leading 0 byte followed by data bytes. This
+		 * array format is used to initialize BigInteger with logical unsigned
+		 * value.  
+		 * 2022-02-21 dsh issue 300 support unsigned and signed 2s compliment negative numbers with leading x'ff' vs x'00'
```
Comment (jyganci) [inline src/pz390.java:13952]: I think you added this comment when you were first addressing the issue about signed numbers. It should be deleted.

File: src/pz390.java [hunk 8]
```diff
@@ -13875,14 +13936,28 @@ private int get_long_log_sub_cc() {
...
+		return new_byte;
+	}
+	private byte[] get_signed_bytes(byte[] data_byte, int data_offset, int data_len) {  // #300
 		/*
 		 * return byte array with leading 0 byte followed by data bytes. This
 		 * array format is used to initialize BigInteger with logical unsigned
-		 * value.
+		 * value.  
+		 * 2022-02-21 dsh issue 300 fix to support 2s compliment negative numbers with leading x'ff' vs x'00'
 		 */
```
Comment (jyganci) [inline src/pz390.java]: The "/* ... */" comment is the same as the one in get_unsigned_bytes(). Change the text to read something like

"Return signed byte array. The returned value is used to initialize a BigInteger with a signed value."

File: src/pz390.java [hunk 9]
```diff
@@ -13875,14 +13936,28 @@ private int get_long_log_sub_cc() {
...
 		byte[] new_byte = new byte[data_len + 1];
+		if (data_byte[data_offset] < 0) {
+			new_byte[0] = (byte)0xff; // issue 300 support 2s compliment negative integers
+		}
 		System.arraycopy(data_byte, data_offset, new_byte, 1, data_len); // RPI
```
Comment (jyganci) [inline src/pz390.java]: Since data_byte[data_offset] is already signed, I think the code could be simplified to
```
    byte[] new_byte = new byte[data_len];
    System.arraycopy(data_byte, data_offset, new_byte, 0, data_len);
```

**Comment by dsh33782** [comment]
John, all

 

I agree, I will update testins5.mlc to include resetting cc for each test.

 

Don Higgins

 ***@***.***> ***@***.***

www.don-higgins.net

File: src/pz390.java [hunk 10]
```diff
@@ -8129,12 +8130,34 @@ private void ins_B9XX(){
...
+			big_int2 = new BigInteger(get_signed_bytes(reg_byte, rf3, 8));
+			big_int1 = big_int1
+			         .multiply(big_int2);
+					 if (big_int1.compareTo(bi_max_pos_long) == 1
+					     || big_int1.compareTo(bi_min_neg_long) == -1) {                  
+				psw_cc = psw_cc3;                          
+				break;                                     
+			} 
+			fp_bi_to_wreg(reg_byte,rf1,big_int1, 8);
+	    	break;
```
Comment (jyganci) [inline src/pz390.java]: I thought I added comments here a few days ago. They're gone??? Repeat them now.

MSGRKC sets CC 0,1,2,3. Only CC 3 set in above code.

MSGRKC (and MSC, MSGC, MSRKC) can result in a fixed-point overflow exception if corresponding PSW Program Mask bit is set. See code for AR instruction on how CC is set, accounting for this possibility. MSC, MSGC, MSGRKC, and MSRKC should do something similar.
